### ISSUE TITLE
An ambiguous situation with the cast of OBJECT and PSZ types.

### ISSUE THREAD

**Issue by leon-ts** [X-Sharp/XSharpPublic#415]
When I debugging my program, I found an ambiguous situation with the conversion of the OBJECT type to PSZ and vice versa.
```
LOCAL o AS OBJECT
LOCAL p AS PSZ
o := StringAlloc("abc")
p := (PSZ)o
```
The variable 'o' contains the correct value "abc" and the type PSZ. But the variable 'p' contains a set of some characters that are not equal to the string "abc".

Maybe I need to use some other type conversion? Or is it a bug?

**Comment by cpyrgas** [comment]
Yeah, the above prints something like "hWt" in my machine. Not sure if there's a feasible way to make this work as expected, I tried a couple things and did not work, maybe Robert has an idea.

But in any case, since those are VO-specific data types, I would also use only what works in VO (the above reports an error in VO that conversion from PSZ to OBJECT is not allowed) as well, so I would type my "o" variable as USUAL and that works. fine 

But I agree, it's not good allowing the conversion to OBJECT and have it silently return a random result. This either needs to become a compiler error, or it must work as when typing the var AS USUAL

**Comment by leon-ts** [comment]
Chris,
In VO, the OBJECT type is not the same as in NET. In VO, this type is for reference types only. And in NET, the OBJECT type is the base type for everything. Well, you already know this better than me :) My path to NET is just beginning.

As for the described problem, this type conversion in my code was a consequence of the fact that earlier in RDD in VO values ​​came as USUAL (_ITEM structure), and in IRdd in XSharp value types come as OBJECT. Where I use PSZ, so as not to do a lot of internal intermediate type conversions, I tried to perform the operation described above (PSZ -> OBJECT -> PSZ), but this did not work. I use PSZ to store an array of strings in memory, where my RDD is required to have the highest possible performance when working with strings. Therefore, in these places of the program I can’t refuse PSZ. And taking into account the fact that IRdd passes parameters as an OBJECT type, I would like to have such a conversion.

**Comment by cpyrgas** [comment]
Leonid,
The important difference is that USUAL is a X# specific custom defined type, so we can do absolutely anything we want to do with it regarding conversions etc, while System.Object is a backed in the system type that we cannot touch, can only add some compiler and runtime trickery to adjust some of its behavior. We'll look into this, but I still think the bet solution is to just use USUAL as an interim variable.

**Comment by RobertvanderHulst** [comment]
I checked and this code boxes the Psz() in the line 
`o := StringAlloc("abc")`

This means that the structure gets boxed, and the only element in the structure is the pointer to the data.

but the in the line
`p := (PSZ) o`
it does not unbox the PSZ but it assumes that the object contains a pointer and calls the PSZ constructor with that pointer.

Mmm. there is something going wrong at a very low level.

This works, because of a recent change I made to box and unbox IntPtr into an object:

```
FUNCTION Start  AS VOID
LOCAL o AS OBJECT
LOCAL p AS PSZ
o := (IntPtr) StringAlloc("abc")              // the intptr gets boxed
p := (PSZ) (IntPtr) o                         // and unboxed again
? p
RETURN

```
I guess I will have to do something similar for PSZ values.

Robert
